### PR TITLE
feat: Upgrade to SDK v4

### PR DESCRIPTION
**Pull request by maaarcelino**
Hi there! This is a PR to update to our latest SDK! We are also rolling out hub shortly - https://hub.cloudquery.io so, once this PR is approved we can follow-up with additional PR so publish the plugin to our hub so users can enjoy dynamic searchable documentation for all versions and all tables.

### CODE REVIEW

File: plugin/plugin.go [hunk 1]
```diff
@@ -1,25 +1,22 @@
...
-	"github.com/coinpaprika/cq-source-coinpaprika/client"
-	"github.com/coinpaprika/cq-source-coinpaprika/resources/services/coins"
-	"github.com/coinpaprika/cq-source-coinpaprika/resources/services/exchanges"
+	"github.com/cloudquery/plugin-sdk/v4/plugin"
 )
 
 var (
+	Kind    = "source"
+	Team    = "coinpaprika"
+	Name    = "coinpaprika-coinpaprika"
```
Comment (candiduslynx) [inline plugin/plugin.go:10]:
```suggestion
	Name    = "coinpaprika"
```

File: resources/services/coins/tickers.go [hunk 2]
```diff
@@ -52,7 +52,7 @@ func fetchTickers(ctx context.Context, meta schema.ClientMeta, parent *schema.Re
 	startDate := cl.StartDate
 	key := fmt.Sprintf(stateKeyTpl, *c.ID)
 	if cl.Backend != nil {
-		value, err := cl.Backend.Get(ctx, key, cl.ID())
+		value, err := cl.Backend.GetKey(ctx, key)
```
Comment (candiduslynx) [inline resources/services/coins/tickers.go:55]: IMO you could just use state.NopClient & never perform the `if cl.Backend != nil` check

File: resources/services/coins/tickers.go [hunk 3]
```diff
@@ -93,9 +93,13 @@ func fetchTickers(ctx context.Context, meta schema.ClientMeta, parent *schema.Re
 		res <- tt
 	}
 	if cl.Backend != nil {
-		err = cl.Backend.Set(ctx, key, cl.ID(), upTo.Format(time.RFC3339))
+		err = cl.Backend.SetKey(ctx, key, upTo.Format(time.RFC3339))
 		if err != nil {
-			return fmt.Errorf("set state failure: %w", err)
+			return fmt.Errorf("failed to save state to backend: %w", err)
+		}
+		err = cl.Backend.Flush(ctx)
```
Comment (candiduslynx) [inline resources/services/coins/tickers.go:100]: no need to perform manual flush, done in `plugin.Client.Sync`

File: plugin/client.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+const maxMsgSize = 100 * 1024 * 1024 // 100 MiB
+
+type Client struct {
+	logger      zerolog.Logger
+	config      client.Spec
+	backendConn *grpc.ClientConn
```
Comment (candiduslynx) [inline plugin/client.go:26]:
```suggestion
```

File: plugin/client.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	scheduler   *scheduler.Scheduler
+	tables      schema.Tables
+	plugin.UnimplementedDestination
+}
+
+func (c *Client) Close(_ context.Context) error {
+	if c.backendConn != nil {
+		return c.backendConn.Close()
+	}
```
Comment (candiduslynx) [inline plugin/client.go:35]:
```suggestion
```

File: plugin/client.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+func (c *Client) Tables(_ context.Context, options plugin.TableOptions) (schema.Tables, error) {
+	tt, err := c.tables.FilterDfs(options.Tables, options.SkipTables, options.SkipDependentTables)
+	if err != nil {
+		return nil, err
+	}
+	return tt, nil
+}
+
+func (c *Client) Sync(ctx context.Context, options plugin.SyncOptions, res chan<- message.SyncMessage) error {
+	tt, err := c.tables.FilterDfs(options.Tables, options.SkipTables, options.SkipDependentTables)
```
Comment (candiduslynx) [inline plugin/client.go:48]: verify for no connection, see https://github.com/cloudquery/cloudquery/blob/main/plugins/source/gcp/resources/plugin/client.go

File: plugin/client.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+		)
+		if err != nil {
+			return fmt.Errorf("failed to dial grpc source plugin at %s: %w", options.BackendOptions.Connection, err)
+		}
+		stateClient, err = state.NewClient(ctx, c.backendConn, options.BackendOptions.TableName)
+		if err != nil {
+			return fmt.Errorf("failed to create state client: %w", err)
+		}
+		c.logger.Info().Str("table_name", options.BackendOptions.TableName).Msg("Connected to state backend")
+	}
```
Comment (candiduslynx) [inline plugin/client.go:74]:
```suggestion
	stateClient := state.Client(&state.NoOpClient{})
	if options.BackendOptions != nil {
		conn, err := grpc.DialContext(ctx, options.BackendOptions.Connection,
			grpc.WithTransportCredentials(insecure.NewCredentials()),
			grpc.WithDefaultCallOptions(
				grpc.MaxCallRecvMsgSize(maxMsgSize),
				grpc.MaxCallSendMsgSize(maxMsgSize),
			),
		)
		if err != nil {
			return fmt.Errorf("failed to dial grpc destination plugin at %s: %w", options.BackendOptions.Connection, err)
		}
		stateClient, err = state.NewClient(ctx, conn, options.BackendOptions.TableName)
		if err != nil {
			return fmt.Errorf("failed to create state client: %w", err)
		}
	}
```

**Review by candiduslynx — CHANGES_REQUESTED** [review]

File: plugin/client.go (hunk 4 above)
Comment (maaarcelino) [reply plugin/client.go]: Done in f40bcbd

File: plugin/client.go (hunk 5 above)
Comment (maaarcelino) [reply plugin/client.go]: Done in f40bcbd

File: plugin/client.go (hunk 7 above)
Comment (maaarcelino) [reply plugin/client.go]: Done in f40bcbd

File: resources/services/coins/tickers.go (hunk 2 above)
Comment (candiduslynx) [reply resources/services/coins/tickers.go:55]: still here

File: resources/services/coins/tickers.go [hunk 8]
```diff
@@ -93,9 +93,9 @@ func fetchTickers(ctx context.Context, meta schema.ClientMeta, parent *schema.Re
 		res <- tt
 	}
 	if cl.Backend != nil {
-		err = cl.Backend.Set(ctx, key, cl.ID(), upTo.Format(time.RFC3339))
+		err = cl.Backend.SetKey(ctx, key, upTo.Format(time.RFC3339))
```
Comment (candiduslynx) [inline resources/services/coins/tickers.go:96]: remove `if cl.Backend != nil` check

**Review by candiduslynx — CHANGES_REQUESTED** [review]

**Review by candiduslynx — APPROVED** [review]

**Comment by candiduslynx** [comment]
cc: @mbillewicz @r--w

**Comment by r--w** [comment]
Wow, thanks for this contribution; it seems like you spent some time on this PR. We will review the code within a week.

**Comment by r--w** [comment]
Hey @candiduslynx & @maaarcelino could you fix the [quality/linter issues](https://github.com/coinpaprika/cq-source-coinpaprika/actions/runs/6721680700/job/18301755791?pr=7#step:4:25)?

You can run this command locally `make check`

**Comment by maaarcelino** [comment]
@r--w There's only a `make lint` command which passes for me locally without making any changes. Those logs appear to point to some other code (e.g. `deleteValue` is not present in any of the files in this repo). 

There were a few failing tests due to a nil pointer and I've fixed those. Now both `make lint` and `make test` all pass.

**Comment by mbillewicz** [comment]
Try to update linter in .github/workflows/test.yaml line 26 to latest version.

**Comment by maaarcelino** [comment]
@mbillewicz Updated to 1.55.1

**Comment by mbillewicz** [comment]
@maaarcelino  pls also regenerate files by  `make gen` due to failing workflow test job

**Pull request merged**